toolchain/lex/tokenized_buffer.cpp

```diff
@@ -839,50 +857,56 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
   CARBON_DISPATCH_LEX_TOKEN(LexStringLiteral)
 
   // A custom dispatch functions that pre-select the symbol token to lex.
-#define CARBON_DISPATCH_LEX_SYMBOL_TOKEN(LexMethod)                          \
-  static auto Dispatch##LexMethod##SymbolToken(Lexer& lexer,                 \
-                                               llvm::StringRef& source_text) \
-      ->void {                                                               \
-    LexResult result = lexer.LexMethod##SymbolToken(                         \
-        source_text, OneCharTokenKindTable[source_text.front()]);            \
-    CARBON_CHECK(result) << "Failed to form a token!";                       \
-    [[clang::musttail]] return DispatchNext(lexer, source_text);             \
+#define CARBON_DISPATCH_LEX_SYMBOL_TOKEN(LexMethod)                           \
+  static auto Dispatch##LexMethod##SymbolToken(                               \
+      Lexer& lexer, llvm::StringRef source_text, ssize_t position)            \
+      ->void {                                                                \
+    LexResult result = lexer.LexMethod##SymbolToken(                          \
+        source_text, OneCharTokenKindTable[source_text[position]], position); \
+    CARBON_CHECK(result) << "Failed to form a token!";                        \
+    [[clang::musttail]] return DispatchNext(lexer, source_text, position);    \
   }
   CARBON_DISPATCH_LEX_SYMBOL_TOKEN(LexOneChar)
   CARBON_DISPATCH_LEX_SYMBOL_TOKEN(LexOpening)
   CARBON_DISPATCH_LEX_SYMBOL_TOKEN(LexClosing)
 
   // Define a set of non-token dispatch functions that handle things like
   // whitespace and comments.
-#define CARBON_DISPATCH_LEX_NON_TOKEN(LexMethod)                              \
-  static auto Dispatch##LexMethod(Lexer& lexer, llvm::StringRef& source_text) \
-      ->void {                                                                \
-    lexer.LexMethod(source_text);                                             \
-    [[clang::musttail]] return DispatchNext(lexer, source_text);              \
+#define CARBON_DISPATCH_LEX_NON_TOKEN(LexMethod)                             \
+  static auto Dispatch##LexMethod(Lexer& lexer, llvm::StringRef source_text, \
+                                  ssize_t position)                          \
+      ->void {                                                               \
+    lexer.LexMethod(source_text, position);                                  \
```

Comment:
I wonder if we'd generate better code if we passed and returned `position` by value in `LexMethod`s, instead of passing it by mutable reference, to avoid any risk the optimizer might think it aliases the source text.

Reply:
All of the `LexMethod`s get inlined away, and so it *shouldn't* be a problem. I can run an experiment to see if there are further improvements though. And there is at least one weird instruction that I noticed that might just be because of this.

If it's OK, I'll run that experiment as a follow-up?

Reply:
Definitely OK to investigate this as a follow-up.